iroh-ffi/src/lib.rs

```diff
@@ -0,0 +1,279 @@
+use std::{path::PathBuf, str::FromStr};
+
+use ::safer_ffi::prelude::*;
+use anyhow::{Context, Result};
+use indicatif::{HumanBytes, HumanDuration};
+use tokio::io::AsyncWriteExt;
+use tokio::runtime::Runtime;
+
+use crate::util::Blake3Cid;
+use iroh::blobs::{Blob, Collection};
+use iroh::get::get_response_machine::{ConnectedNext, EndBlobNext};
+use iroh::get::{self, get_data_path, get_missing_range, get_missing_ranges, pathbuf_from_name};
+use iroh::protocol::GetRequest;
+use iroh::provider::Ticket;
+use iroh::tokio_util::SeekOptimized;
+use iroh::{Hash, PeerId};
+
+mod util;
+
+const MAX_CONCURRENT_DIALS: u8 = 16;
+
+#[ffi_export]
+fn iroh_get(
+    hash: char_p::Ref<'_>,
+    peer: char_p::Ref<'_>,
+    peer_addr: char_p::Ref<'_>,
+    out_path: char_p::Ref<'_>,
+) -> u32 {
+    let result = std::panic::catch_unwind(|| {
+        let hash = hash.to_str().parse::<Hash>().unwrap();
+        let peer = peer.to_str().parse::<PeerId>().unwrap();
+        let peer_addr = peer_addr.to_str().parse().unwrap();
+        let out_path = PathBuf::from_str(out_path.to_str()).unwrap();
+        let rt = Runtime::new().unwrap();
```

Comment:
this is a lot of `unwrap`, please use `anyhow::Result` to not just run into panics unnecessarily